openeo/extra/job_management.py

```diff
@@ -0,0 +1,290 @@
+    """
+    Tracker for multiple jobs on multiple backends.
+
+    Usage example:
+
+        manager = MultiBackendJobManager()
+        manager.add_backend("foo", connection=openeo.connect("http://foo.test"))
+        manager.add_backend("bar", connection=openeo.connect("http://bar.test"))
+
+        jobs_df = pd.DataFrame(....)
+        output_file = Path("jobs.csv")
+        def start_job(row, connection, **kwargs):
+            ...
+
+        manager.run_jobs(df=df, start_job=start_job, output_file=output_file)
+
+    """
+
+    def __init__(self, poll_sleep: int = 60):
+        """Create a MultiBackendJobManager.
+
+        :param poll_sleep: How many seconds to sleep between polls.
+        """
+        self.backends: Dict[str, _Backend] = {}
+        self.poll_sleep = poll_sleep
+
+    def add_backend(
+        self,
+        name: str,
+        connection: Union[Connection, Callable[[], Connection]],
+        parallel_jobs: int = 2,
+    ):
+        """
+        Register a backend with a name and a Connection getter.
+
+        :param name:
+            Name of the backend.
+        :param connection:
+            Either a Connection to the backend, or a callable to create a backend connection.
+        :param parallel_jobs:
+            Maximum number of jobs to allow in parallel on a backend.
+        """
+        if isinstance(connection, Connection):
+            c = connection
+            connection = lambda: c
+        assert callable(connection)
+        self.backends[name] = _Backend(
+            get_connection=connection, parallel_jobs=parallel_jobs
+        )
+
+    def _normalize_df(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Ensure we have the required columns and the expected type for the geometry column.
+
+        :param df: The dataframe to normalize.
+        :return: a new dataframe that is normalized.
+        """
+
+        # check for some required columns.
+        required_with_default = [
+            ("status", "not_started"),
+            ("id", None),
+            ("start_time", None),
+            ("cpu", None),
+            ("memory", None),
+            ("duration", None),
+            ("backend_name", None),
+        ]
+        new_columns = {
+            col: val for (col, val) in required_with_default if col not in df.columns
+        }
+        df = df.assign(**new_columns)
+        # Workaround for loading of geopandas "geometry" column.
+        if "geometry" in df.columns and df["geometry"].dtype.name != "geometry":
+            df["geometry"] = df["geometry"].apply(shapely.wkt.loads)
+        return df
+
+    # TODO: long method with deep nesting. Refactor it to make it more readable.
+    def run_jobs(self, df: pd.DataFrame, start_job: Callable, output_file: Path):
+        """Runs jobs, specified in a dataframe, and tracks parameters.
+
+        :param df:
+            DataFrame that specifies the jobs, and tracks the jobs' statuses.
+        :param start_job:
```

Comment:
From reading the implementation it seems that `start_job` should actually be called `create_job`, right @jdries ?

also the documentation should state that this callable is expected to return a `openeo.rest.job.BatchJob` object

Reply:
nevermind, the start_job callable is apparently also allowed to start the job already

Reply:
Added to docstring that startt_job should return openeo.rest.job.BatchJob and updated the type hint accordingly.